Snapshotting unittest.mock.call objects fails

Attempting to snapshot the `call_args_list` of a mock fails with this stack trace:

```
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/module.py:261: in assert_match
    self.assert_value_matches_snapshot(value, prev_snapshot)
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/module.py:241: in assert_value_matches_snapshot
    formatter.assert_value_matches_snapshot(self, test_value, snapshot_value, Formatter())
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/formatters.py:21: in assert_value_matches_snapshot
    test.assert_equals(formatter.normalize(test_value), snapshot_value)
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/formatter.py:24: in normalize
    return formatter.normalize(value, self)
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/formatters.py:45: in normalize
    return value.__class__(formatter.normalize(item) for item in iterator)
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/formatters.py:45: in <genexpr>
    return value.__class__(formatter.normalize(item) for item in iterator)
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/formatter.py:24: in normalize
    return formatter.normalize(value, self)
.wavemodem-venv/lib/python3.7/site-packages/snapshottest/formatters.py:45: in normalize
    return value.__class__(formatter.normalize(item) for item in iterator)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'unittest.mock._Call'>
value = <generator object CollectionFormatter.normalize.<locals>.<genexpr> at 0x7f17dabc5050>, name = '', parent = None
two = False, from_kall = True

    def __new__(cls, value=(), name='', parent=None, two=False,
                from_kall=True):
        args = ()
        kwargs = {}
>       _len = len(value)
E       TypeError: object of type 'generator' has no len()

../../../.pyenv/versions/3.7.6/lib/python3.7/unittest/mock.py:2052: TypeError
```

PR to fix it incoming :slightly_smiling_face:

--- a/snapshottest/formatters.py
+++ b/snapshottest/formatters.py
@@ -40,7 +40,9 @@ class TypeFormatter(BaseFormatter):
 class CollectionFormatter(TypeFormatter):
     def normalize(self, value, formatter):
         iterator = iter(value.items()) if isinstance(value, dict) else iter(value)
-        return value.__class__(formatter.normalize(item) for item in iterator)
+        # Normally we shouldn't need to turn this into a list, but some iterable
+        # constructors need a list not an iterator (e.g. unittest.mock.call).
+        return value.__class__([formatter.normalize(item) for item in iterator])
 
 
 class DefaultDictFormatter(TypeFormatter):

--- a/snapshottest/formatters.py
+++ b/snapshottest/formatters.py
@@ -40,6 +40,7 @@ class TypeFormatter(BaseFormatter):
 class CollectionFormatter(TypeFormatter):
     def normalize(self, value, formatter):
         iterator = iter(value.items()) if isinstance(value, dict) else iter(value)
+        # https://github.com/syrusakbary/snapshottest/issues/115
         # Normally we shouldn't need to turn this into a list, but some iterable
         # constructors need a list not an iterator (e.g. unittest.mock.call).
         return value.__class__([formatter.normalize(item) for item in iterator])

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -6,6 +6,9 @@ import six
 
 from snapshottest.formatter import Formatter
 
+if not six.PY2:
+    import unittest.mock
+
 
 @pytest.mark.parametrize("text_value, expected", [
     # basics
@@ -51,3 +54,9 @@ def test_non_ascii_text_formatting(text_value, expected_py3, expected_py2):
     formatter = Formatter()
     formatted = formatter(text_value)
     assert formatted == expected
+
+
+if not six.PY2:
+    def test_can_normalize_unittest_mock_call_object():
+        formatter = Formatter()
+        print(formatter.normalize(unittest.mock.call(1, 2, 3)))

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -57,6 +57,7 @@ def test_non_ascii_text_formatting(text_value, expected_py3, expected_py2):
 
 
 if not six.PY2:
+    # See issue #115
     def test_can_normalize_unittest_mock_call_object():
         formatter = Formatter()
         print(formatter.normalize(unittest.mock.call(1, 2, 3)))

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -57,7 +57,7 @@ def test_non_ascii_text_formatting(text_value, expected_py3, expected_py2):
 
 
 if not six.PY2:
-    # See issue #115
+    # https://github.com/syrusakbary/snapshottest/issues/115
     def test_can_normalize_unittest_mock_call_object():
         formatter = Formatter()
         print(formatter.normalize(unittest.mock.call(1, 2, 3)))

--- a/tests/test_formatter.py
+++ b/tests/test_formatter.py
@@ -60,3 +60,7 @@ if not six.PY2:
     def test_can_normalize_unittest_mock_call_object():
         formatter = Formatter()
         print(formatter.normalize(unittest.mock.call(1, 2, 3)))
+
+    def test_can_normalize_iterator_objects():
+        formatter = Formatter()
+        print(formatter.normalize(x for x in range(3)))

